enketo/enketo-express
# Upgrade Ubuntu version in CI, Node/NPM dependencies, fix audit vulnerabilities

### Changes

- Upgraded Ubuntu version in CI:
  - This unblocks other relevant updates
  - It also unblocks testing in an up to date version of Chrome (previously limited to a version around 40 releases behind)
  - Because the Ubuntu update replaces Upstart (its previous supervisor) with Systemd, I first tried to convert the Redis service to Systemd. This proved to be a huge time sink, and we decided instead that it's okay to start the Redis directly services without supervision.
- Upgraded `enketo-transformer` ~~temporarily to `https://github.com/eyelidlessness/enketo-transformer.git#node-16-dependency-upgrades`, this change will depend on a transformer release~~
- Upgraded `enketo-core` ~~temporarily to `https://github.com/eyelidlessness/enketo-core.git#node-16-dependency-upgrades`, this change will also depend on a core release~~
- Pin engines to Node >=14 <17, NPM >=6 <7
- Pin Volta to Node 16/NPM 6
- Set up matrix testing for Travis CI to use NPM 14 and 16
- Added `preinstall` script to detect when local install includes `libxslt@0.9.0`, if so run `npm ci` to ensure all previously compiled binaries are upgraded for the current Node version
- Upgraded [sub-]dependencies for outstanding security vulnerabilities where possible
- `npm update` for dependencies with semver-compatible upgrades
- Upgraded other outdated dependencies
- Removed broken code coverage badge (see enketo/enketo-core#795; in express the number doesn't render 100% because coverage of the backend is still working, but coverage of the frontend breaks in the same way as in enketo-core)

### Validation steps taken

- `npm clean-install`
- `npm install` with previous `node_modules` in place
- `npm audit` shows only vulnerabilities which cannot currently be patched, which are:
  - `enketo-core >mergexml > xmldom`: the dependency is `0.y.z` and the fix is treated as semver-incompatible
- `npm outdated` shows only false positives:
  - `vex-dialog-enketo`
- `npm run test` passes (this also includes the `eslint:check` Grunt task)
  - Two tests were updated in `media-controller.spec.js`. In both cases, they correctly expected the action to produce an error, but the error message has changed. I strongly suspect this is a change in Node's internal network behavior, as the previously installed version of the relevant dependency (`request-filtering-agent`) produces the same behavior in Node 16
- `npx grunt develop`, then:
  - submit a form online
  - save a draft offline
  - submit a form offline
  - submit an edited draft offline
  - submit an edit
  - preview a form
- `npm run build-docs` before/after doesn't produce a diff, other than those expected from substantive changes:
  - `docs/index.html` reflects removal of code coverage badge
- `npx grunt` (build step) does produce a diff, but that appears to be mainly a product of:
  - Upgrading `esbuild`, it moves some stuff around presumably for efficiency
  - Upgrading other packages that are bundled with the build:
    - Substantial changes to `jszip`
    - Moderate changes to `i18next`
    - Change to enketo-core's `Form.requiredTransformerVersion`
